### PR TITLE
Use dotnet pack for other Microsoft.extensions* projects

### PR DESCRIPTION
**Pull request by Anipik**
Related to https://github.com/dotnet/runtime/pull/51765

- removes pkgprojs
- enable using dotnet pack task

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Logging/src/Microsoft.Extensions.Logging.csproj [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
   <PropertyGroup>
     <TargetFrameworks>netstandard2.1;netstandard2.0;net461</TargetFrameworks>
+    <IsPackable>false</IsPackable>
```
Comment (safern) [inline src/libraries/Microsoft.Extensions.Logging/src/Microsoft.Extensions.Logging.csproj:5]: Why Logging is not packable? You are porting this one in a separate PR?

Comment (Anipik) [reply src/libraries/Microsoft.Extensions.Logging/src/Microsoft.Extensions.Logging.csproj:5]: This one contains source generator, so we will probably needs some hacking and workarounds for making it work

Comment (ViktorHofer) [reply src/libraries/Microsoft.Extensions.Logging/src/Microsoft.Extensions.Logging.csproj:5]: @Anipik for cases where you would expect a change to not be self-speaking could you please either add a comment inline into the sources (in this case i.e. `TODO: Figure out how to pack source generators`) or in the PR? This will help your reviewers and yourself to get better feedback and avoid unnecessary review iterations.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/Microsoft.Extensions.HostFactoryResolver.Sources.csproj [hunk 2]
```diff
@@ -4,5 +4,11 @@
     <TargetFrameworks>netstandard2.0</TargetFrameworks>
     <!-- Avoid trying to binplace non existent documentation file.-->
     <EnableBinPlacing>false</EnableBinPlacing>
+    <IsShipping>false</IsShipping>
```
Comment (ViktorHofer) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/Microsoft.Extensions.HostFactoryResolver.Sources.csproj:7]: I think the comment in L2 isn't valid anymore. (can't comment on it in the PR, which is why I chose this line)

File: src/libraries/Directory.Build.props [hunk 3]
```diff
@@ -99,6 +99,7 @@
     <StrongNameKeyId>Open</StrongNameKeyId>
     <!-- Microsoft.Extensions projects have a separate StrongNameKeyId -->
     <StrongNameKeyId Condition="$(MSBuildProjectName.StartsWith('Microsoft.Extensions.'))">MicrosoftAspNetCore</StrongNameKeyId>
+    <IsPackable Condition="$(MSBuildProjectName.StartsWith('Microsoft.Extensions.'))">true</IsPackable>
```
Comment (ViktorHofer) [inline src/libraries/Directory.Build.props:102]:
```suggestion
    <!-- TODO: Remove condition when all libraries migrated from pkgprojs to NuGet's Pack Task. -->
    <IsPackable Condition="$(MSBuildProjectName.StartsWith('Microsoft.Extensions.'))">true</IsPackable>
```

**Comment by ViktorHofer** [comment]
@Anipik please add the nuspec diff for all the packages that you are touching as part of this PR.

**Comment by Anipik** [comment]
> @Anipik please add the nuspec diff for all the packages that you are touching as part of this PR.

do u want me to attach all the diffs ? most of them are non-interseting and similar to each other

**Comment by ViktorHofer** [comment]
> do u want me to attach all the diffs ? most of them are non-interseting and similar to each other

They don't need to inlined as a comment on GH, uploading them in a zip would suffice as well. Usually for minor packaging changes we wouldn't be interested in the nuspec diff but for such a fundamental change a diff makes sense. See it as an additional safeguard to avoid an unexpected regression. It might also help people who come back to this PR in the future to better understand the change.

**Comment by Anipik** [comment]
https://gist.github.com/Anipik/712f51a231dd08899dc4def131933743

here is the diff for all the extensions nuspec.

**Comment by Anipik** [comment]
@ViktorHofer @safern @ericstj can you review this one ?

**Comment by ViktorHofer** [comment]
Thanks for the diff. Here are the things that stand out:
1. `<licenseUrl>https://licenses.nuget.org/MIT</licenseUrl>`. LicenseUrl (and the msbuild property `PackageLicenseUrl`) is deprecated. From a search I couldn't find the place that sets it.
2. Please again move the `PackageDescription` properties from the Directory.Build.props file into the src project.
3. Why are `build` and `analyzer` items excluded? `exclude="Build,Analyzers"`
4. Is it expected that the xml files are included from the package cache? `.nuget\packages\microsoft.private.intellisense\5.0.0-preview-20201009.2\IntellisenseFiles\net\1033\Microsoft.Extensions.Caching.Abstractions.xml`
5. Do we still want/need the `servicable` element? It's currently removed.

I recommend that we do another review of the produced diff after these points are discussed / addressed.

**Comment by ericstj** [comment]
> Why are build and analyzer items excluded? exclude="Build,Analyzers"

That's the default behavior of Pack.  I agree it's a bit annoying.  I'm OK if we remove it from our packages, or leave as is to be more "normal".

> Is it expected that the xml files are included from the package cache? .nuget\packages\microsoft.private.intellisense\5.0.0-preview-20201009.2\IntellisenseFiles\net\1033\Microsoft.Extensions.Caching.Abstractions.xml

CC @safern @carlossanlop.  I believe we need to have some form of state that drives this.  If a library has done the work to move intellisense to triple-slash then that should be used (new model).  If not, then the microsoft.private.intellisense package should be used (old model).  So long as we aren't changing it with this PR it is OK.

> Do we still want/need the servicable element? It's currently removed.

Yes we want it.  It's what causes .NETCore to write breadcrumbs for servicing of OOBs.

**Comment by Anipik** [comment]
> Do we still want/need the servicable element? It's currently removed.

dotnet pack add this to the symbols nuspec file and set it to true for the symbols package.

sample one.

```xml
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2012/06/nuspec.xsd">
  <metadata>
    <id>Microsoft.Extensions.Caching.Abstractions</id>
    <version>6.0.0-dev</version>
    <authors>Microsoft</authors>
    <license type="expression">MIT</license>
    <licenseUrl>https://licenses.nuget.org/MIT</licenseUrl>
    <icon>Icon.png</icon>
    <projectUrl>https://dot.net/</projectUrl>
    <description>Caching abstractions for in-memory cache and distributed cache.

Commonly Used Types:
Microsoft.Extensions.Caching.Distributed.IDistributedCache
Microsoft.Extensions.Caching.Memory.IMemoryCache</description>
    <releaseNotes>https://go.microsoft.com/fwlink/?LinkID=799421</releaseNotes>
    <copyright>© Microsoft Corporation. All rights reserved.</copyright>
    <serviceable>true</serviceable>
    <repository type="git" url="https://github.com/dotnet/runtime" commit="0000000000000000000000000000000000000000" />
    <dependencies>
      <group targetFramework=".NETFramework4.6.1">
        <dependency id="Microsoft.Extensions.Primitives" version="6.0.0-dev" exclude="Build,Analyzers" />
      </group>
      <group targetFramework=".NETStandard2.0">
        <dependency id="Microsoft.Extensions.Primitives" version="6.0.0-dev" exclude="Build,Analyzers" />
      </group>
    </dependencies>
  </metadata>
  <files>
    <file src="C:\git\runtime4\artifacts\bin\Microsoft.Extensions.Caching.Abstractions\net461-Debug\Microsoft.Extensions.Caching.Abstractions.pdb" target="lib\net461\Microsoft.Extensions.Caching.Abstractions.pdb" />
    <file src="C:\git\runtime4\artifacts\bin\Microsoft.Extensions.Caching.Abstractions\netstandard2.0-Debug\Microsoft.Extensions.Caching.Abstractions.pdb" target="lib\netstandard2.0\Microsoft.Extensions.Caching.Abstractions.pdb" />
    <file src="C:\git\runtime4\artifacts\bin\Microsoft.Extensions.Caching.Abstractions\net461-Debug\Microsoft.Extensions.Caching.Abstractions.dll" target="lib\net461\Microsoft.Extensions.Caching.Abstractions.dll" />
    <file src="C:\git\runtime4\artifacts\bin\Microsoft.Extensions.Caching.Abstractions\netstandard2.0-Debug\Microsoft.Extensions.Caching.Abstractions.dll" target="lib\netstandard2.0\Microsoft.Extensions.Caching.Abstractions.dll" />
    <file src="C:\Users\anagniho\.nuget\packages\microsoft.private.intellisense\5.0.0-preview-20201009.2\IntellisenseFiles\net\1033\Microsoft.Extensions.Caching.Abstractions.xml" target="lib\net461\Microsoft.Extensions.Caching.Abstractions.xml" />
    <file src="C:\Users\anagniho\.nuget\packages\microsoft.private.intellisense\5.0.0-preview-20201009.2\IntellisenseFiles\net\1033\Microsoft.Extensions.Caching.Abstractions.xml" target="lib\netstandard2.0\Microsoft.Extensions.Caching.Abstractions.xml" />
    <file src="C:\Users\anagniho\.nuget\packages\microsoft.dotnet.arcade.sdk\6.0.0-beta.21222.1\tools\Assets\DotNetPackageIcon.png" target="Icon.png" />
    <file src="C:\git\runtime4\LICENSE.TXT" target="LICENSE.TXT" />
    <file src="C:\git\runtime4\THIRD-PARTY-NOTICES.TXT" target="THIRD-PARTY-NOTICES.TXT" />
  </files>
</package>
```
if we want it to set for non-symbols package we will probably have to ask the nuget team to do this.

**Comment by ViktorHofer** [comment]
@Anipik can't you set `<Serviceable />` which is passed in to the PackTask? https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets#L256

**Comment by Anipik** [comment]
> Is it expected that the xml files are included from the package cache? .nuget\packages\microsoft.private.intellisense\5.0.0-preview-20201009.2\IntellisenseFiles\net\1033\Microsoft.Extensions.Caching.Abstractions.xml

i checked with @safern  and we are currently fine using the xml files from the nuget

> <licenseUrl>https://licenses.nuget.org/MIT</licenseUrl>. LicenseUrl (and the msbuild property PackageLicenseUrl) is deprecated. From a search I couldn't find the place that sets it.

pack task does this, it appends the $(PackageLicenseExpression) to the default url. we will need to use the $(PackageLiscenseFile) property in order to point it to the licensefile in the package https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#packing-a-license-expression-or-a-license-file

> @Anipik can't you set <Serviceable /> which is passed in to the PackTask? https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets#L256

we are already setting that property to true

**Comment by ViktorHofer** [comment]
> @Anipik can't you set which is passed in to the PackTask? https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Build.Tasks.Pack/NuGet.Build.Tasks.Pack.targets#L256

@dotnet/nuget-team as we want to keep the serviceable attribute on our nuspecs, is there a way to do so via the PackTask? I see the `Serviceable` property being honored but based on @Anipik's diff, the attribute is removed from the non symbol packages.

**Comment by ViktorHofer** [comment]
> pack task does this, it appends the $(PackageLicenseExpression) to the default url. we will need to use the $(PackageLiscenseFile) property in order to point it to the licensefile in the package https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#packing-a-license-expression-or-a-license-file

I just tried this out in a sample app and I can't see a licenseurl being added to the nuspec. Am I missing something?

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <PackageLicenseExpression>MIT</PackageLicenseExpression>
  </PropertyGroup>

</Project>
```

results in this nuspec:

```
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2012/06/nuspec.xsd">
  <metadata>
    <id>licenseurl</id>
    <version>1.0.0</version>
    <authors>licenseurl</authors>
    <owners></owners>
    <requireLicenseAcceptance>false</requireLicenseAcceptance>
    <license type="expression">MIT</license>
    <description>Package Description</description>
    <dependencies>
      <group targetFramework="net6.0" />
    </dependencies>
  </metadata>
</package>
```

**Comment by Anipik** [comment]
I can see the license url
```
<?xml version="1.0" encoding="utf-8"?>
<package xmlns="http://schemas.microsoft.com/packaging/2012/06/nuspec.xsd">
  <metadata>
    <id>ConsoleApp29</id>
    <version>6.0.0</version>
    <authors>ConsoleApp29</authors>
    <license type="expression">MIT</license>
    <licenseUrl>https://licenses.nuget.org/MIT</licenseUrl>
    <description>Package Description</description>
    <dependencies>
      <group targetFramework=".NETCoreApp3.1">
        <dependency id="NuGet.Frameworks" version="5.10.0-preview.2.7185" exclude="Build,Analyzers" />
      </group>
    </dependencies>
  </metadata>
  <files>
    <file src="C:\Users\anagniho\source\repos\ConsoleApp29\bin\Debug\netcoreapp3.1\ConsoleApp29.runtimeconfig.json" target="lib\netcoreapp3.1\ConsoleApp29.runtimeconfig.json" />
    <file src="C:\Users\anagniho\source\repos\ConsoleApp29\bin\Debug\netcoreapp3.1\ConsoleApp29.dll" target="lib\netcoreapp3.1\ConsoleApp29.dll" />
  </files>
</package>
```
can you share the binlog ?

**Comment by ViktorHofer** [comment]
Shared the binlog offline with Anirudh.

**Comment by ericstj** [comment]
> @dotnet/nuget-team as we want to keep the serviceable attribute on our nuspecs, is there a way to do so via the PackTask? I see the Serviceable property being honored but based on @Anipik's diff, the attribute is removed from the non symbol packages.

This didn't repro for me.  Dotnet pack on this gives me a serviceable element in the nuspec. 
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <Serviceable>true</Serviceable>
  </PropertyGroup>
</Project>
```

**Comment by Anipik** [comment]
> I just tried this out in a sample app and I can't see a licenseurl being added to the nuspec. Am I missing something?

Talked with @nkolev92, he pointed out that this is a fallback for older clients (older than dev16) and should not affect anything if we have a license element as well

**Comment by Anipik** [comment]
> This didn't repro for me. Dotnet pack on this gives me a serviceable element in the nuspec.

it could probably linked to my machine because i checked the the package produced in the internal builds and it has servicable=true https://dev.azure.com/dnceng/public/_build/results?buildId=1116205&view=artifacts&pathAsName=false&type=publishedArtifacts

package name = Microsoft.Extensions.Configuration.EnvironmentVariables.

**Comment by ericstj** [comment]
> could probably linked to my machine 

I gave it a try by pulling down your branch and building a couple packages and confirm that serviceable is in the nuspec.

**Comment by ericstj** [comment]
One thing that comes to mind with respect to this effort: a lot of this metadata is validated at publish time.  It might be prudent to do an early dry run to catch anything that might block publishing.  IIRC there is some async validation build that's supposed to catch this: we should look at it.

**Comment by ViktorHofer** [comment]
> One thing that comes to mind with respect to this effort: a lot of this metadata is validated at publish time. It might be prudent to do an early dry run to catch anything that might block publishing. IIRC there is some async validation build that's supposed to catch this: we should look at it.

@ericstj see https://github.com/dotnet/core-eng/pull/12993 which covers how validation issues will be communicated.

**Comment by ViktorHofer** [comment]
@Anipik LGTM in general. Can you please add a fresh diff of the nuspecs with a recent NuGet client so that we can see that the licenseUrl element isn't present but the serviceable element is? Thanks

**Comment by ericstj** [comment]
>  see dotnet/core-eng#12993 which covers how validation issues will be communicated.

Will that be active to catch an issue as soon as this is in main and before we ship preview5?  If not, let's manually check the results once this is in.

File: src/libraries/Microsoft.Extensions.Logging/src/Microsoft.Extensions.Logging.csproj (hunk 1 above)
Comment (ericstj) [reply src/libraries/Microsoft.Extensions.Logging/src/Microsoft.Extensions.Logging.csproj:5]: Here's a sample that demonstrates how to pack an analyzer/source-generator.  No need to tackle in this PR, but might be helpful for a followup: https://github.com/ericstj/runtime/commit/ba5a7c83a5c3c62343f5c285ed52868426c2912e
cc @maryamariyan

**Comment by ViktorHofer** [comment]
> Will that be active to catch an issue as soon as this is in main and before we ship preview5? If not, let's manually check the results once this is in.

Nope, that will go live in June. I agree that we should review the produced packages from an official build post-merge.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/TrimmingTests/Microsoft.Extensions.DependencyInjection.TrimmingTests.proj [hunk 4]
```diff
@@ -3,6 +3,7 @@
 
   <PropertyGroup>
     <AdditionalProjectReferences>Microsoft.Extensions.DependencyInjection</AdditionalProjectReferences>
+    <IsPackable>false</IsPackable>
```
Comment (ViktorHofer) [inline src/libraries/Microsoft.Extensions.DependencyInjection/tests/TrimmingTests/Microsoft.Extensions.DependencyInjection.TrimmingTests.proj:6]: Just curious, why is this necessary?

Comment (Anipik) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/TrimmingTests/Microsoft.Extensions.DependencyInjection.TrimmingTests.proj:6]:
```
  <!-- Remove when https://github.com/NuGet/Home/issues/10405 is implemented and consumed. -->
  <Target Name="ErrorForMissingPackageDescription"
          Condition="'$(IsPackable)' == 'true' and '$(PackageDescription)' == ''">
    <Error Text="Required property 'PackageDescription' is missing for $(MSBuildProjectName)." />
  </Target>

```

This target gets run while running the linker tests and it fails as we moved the package Description to the src project.

File: src/libraries/Directory.Build.props [hunk 5]
```diff
@@ -98,7 +98,9 @@
     <!-- Default any assembly not specifying a key to use the Open Key -->
     <StrongNameKeyId>Open</StrongNameKeyId>
     <!-- Microsoft.Extensions projects have a separate StrongNameKeyId -->
+    <!-- TODO: Remove condition when all libraries migrated from pkgprojs to NuGet's Pack Task. -->
     <StrongNameKeyId Condition="$(MSBuildProjectName.StartsWith('Microsoft.Extensions.'))">MicrosoftAspNetCore</StrongNameKeyId>
+    <IsPackable Condition="$(MSBuildProjectName.StartsWith('Microsoft.Extensions.'))">true</IsPackable>
```
Comment (ViktorHofer) [inline src/libraries/Directory.Build.props:103]: You also want to check on `$(IsSourceProject)` to not mark any other project types as packable. I think that should mitigate the linker issues that you are seeing.

File: src/libraries/Microsoft.Extensions.DependencyInjection/tests/TrimmingTests/Microsoft.Extensions.DependencyInjection.TrimmingTests.proj (hunk 4 above)
Comment (ViktorHofer) [reply src/libraries/Microsoft.Extensions.DependencyInjection/tests/TrimmingTests/Microsoft.Extensions.DependencyInjection.TrimmingTests.proj:6]: See my other comment in the PR. These projects should not be marked as `<IsPackable>true</IsPackable>`.

File: src/libraries/Directory.Build.props [hunk 6]
```diff
@@ -100,9 +100,8 @@
     <!-- Microsoft.Extensions projects have a separate StrongNameKeyId -->
     <!-- TODO: Remove condition when all libraries migrated from pkgprojs to NuGet's Pack Task. -->
     <StrongNameKeyId Condition="$(MSBuildProjectName.StartsWith('Microsoft.Extensions.'))">MicrosoftAspNetCore</StrongNameKeyId>
-    <IsPackable Condition="$(MSBuildProjectName.StartsWith('Microsoft.Extensions.'))">true</IsPackable>
+    <IsPackable Condition="$(MSBuildProjectName.StartsWith('Microsoft.Extensions.')) and '$(IsSourceProject)' != 'true'">true</IsPackable>
```
Comment (ViktorHofer) [inline src/libraries/Directory.Build.props:103]: SourceProject should be true, shouldn't it?

**Comment by safern** [comment]
> CC @safern @carlossanlop. I believe we need to have some form of state that drives this. If a library has done the work to move intellisense to triple-slash then that should be used (new model). If not, then the microsoft.private.intellisense package should be used (old model). So long as we aren't changing it with this PR it is OK.

Yes, this PR is actually making that easier for projects that no longer depend on a pkgproj. 

I would expect we will introduce a new property to specify `<UseDocsFromTripleSlash>` or something like that and if that is set to true, use the generated ones by the compiler rather than the ones from the intellisense package.

**Review by ViktorHofer — APPROVED** [review]

**Comment by ericstj** [comment]
Great to see this merged.  Thanks for the great team work here! 🚀  🚀  🚀

**Comment by adiaaida** [comment]
Notifications won't be live for preview5, likely, but you can monitor Validate-DotNet manually if you need to (Runtime runs: https://dev.azure.com/dnceng/internal/_build?definitionId=827&tagFilter=validation-runtime-.NET%206). We run nuget metadata validation in the NuGet Validation step in the Validation ring. See [here](https://dev.azure.com/dnceng/internal/_build/results?buildId=1126713&view=logs&s=2f4b62b5-8d35-5570-0513-04a0c5199809&j=5bed3f3b-f321-5a90-e93b-0c85e783a76a) for last night's run (it's green).
